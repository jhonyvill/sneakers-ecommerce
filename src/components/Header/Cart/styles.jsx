import styled from "styled-components";

export const Cart = styled.button`
  background-color: transparent;
`;
