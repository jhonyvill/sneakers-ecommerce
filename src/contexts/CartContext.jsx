import React from "react";
import { useReducer } from "react";

export const CartContext = React.createContext();

export const CartProvider = ({ children }) => {
  const initialState = { items: [] };

  function cartReducer(state, action) {
    switch (action.type) {
      case "ADD_ITEM":
        return { ...state, items: [...state.items, action.payload] };
      case "UPDATE_ITEM":
        const { product } = action.payload;
        const itemIndex = state.items.findIndex(
          (item) => item.product.id === product.id
        );

        const updatedItems = [...state.items];
        updatedItems.slice(itemIndex, product);
        return { ...state, items: updatedItems };
      default:
        state;
    }
  }

  const [state, dispatch] = useReducer(cartReducer, initialState);

  return (
    <>
      <CartContext.Provider
        value={{ cartState: state, cartDispatch: dispatch }}
      >
        {children}
      </CartContext.Provider>
    </>
  );
};
