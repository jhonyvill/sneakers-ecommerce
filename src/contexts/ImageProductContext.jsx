import React from "react";
import { useState } from "react";

export const ImageProductContext = React.createContext();

export const ImageProductProvider = ({ children }) => {
  const [selectedImage, setSelectedImage] = useState({
    image: null,
    index: 0,
  });

  const handleSelectedModalImage = (image, index) => {
    setSelectedImage({ image, index });
  };

  return (
    <>
      <ImageProductContext.Provider
        value={{
          imageSelected: selectedImage.image,
          indexImageSelected: selectedImage.index,
          handleSelectedModalImage,
        }}
      >
        {children}
      </ImageProductContext.Provider>
    </>
  );
};
