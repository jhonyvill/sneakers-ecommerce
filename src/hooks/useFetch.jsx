import { useCallback, useState } from "react";

export const useFetch = () => {
  const [data, setData] = useState(null);
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(false);

  const request = useCallback(async (url) => {
    let response;
    let json;
    try {
      setError(null);
      setLoading(true);

      response = await fetch(url);
      json = await response.json();
      if (!response.ok) throw new Error(json.message);
    } catch (error) {
      json = null;
      setError(error.message);
      console.log(error.message);
    } finally {
      setData(json);
      setLoading(false);
    }
  }, []);

  return {
    data,
    error,
    loading,
    request,
  };
};
