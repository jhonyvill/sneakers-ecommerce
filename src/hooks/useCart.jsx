import React from "react";
import { useState, useContext } from "react";
import { CartContext } from "../contexts/CartContext";

export const useCart = () => {
  const { cartState, cartDispatch } = useContext(CartContext);
  const [quantitySelected, setQuantitySelected] = useState(0);
  const cartHasItems = cartState.items.length > 0;

  function findProductInCart(product) {
    return cartState.items.find((item) => item.product.id === product.id);
  }

  function addCartProduct(product) {
    const existentProduct = findProductInCart(product);
    if(existentProduct){
      cartDispatch({
        type: "UPDATE_ITEM",
        payload: {
          product: product,
          quantity: (existentProduct.quantity += quantitySelected),
        },
      });
    }else if(quantitySelected > 0){
      cartDispatch({
        type: "ADD_ITEM",
        payload: { product: product, quantity: quantitySelected },
      });
    }
    setQuantitySelected(0);
  }

  return {
    quantitySelected,
    setQuantitySelected,
    addCartProduct,
    cartHasItems
  };
};
