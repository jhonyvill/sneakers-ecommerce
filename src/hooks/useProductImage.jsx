import React from "react";
import { useContext } from "react";
import { useState } from "react";
import { ImageProductContext } from "../contexts/ImageProductContext";

export const useProductImage = (images) => {
  const { imageSelected, indexImageSelected, handleSelectedModalImage } = useContext(ImageProductContext);
  const [featuredImage, setFeaturedImage] = useState({
    image: imageSelected || images[0],
    index: indexImageSelected,
  });

  function handleChangeImage(image, indexClicked) {
    indexClicked >= 0 &&
      indexClicked <= images.length - 1 &&
      indexClicked !== featuredImage.index &&
      setFeaturedImage({ image, index: indexClicked });
  }
  function handleClickFeaturedImage() {
    handleSelectedModalImage(featuredImage.image, featuredImage.index);
  }

  return {
    featuredImage: featuredImage.image,
    imageIndexActive: featuredImage.index,
    handleChangeImage,
    handleClickFeaturedImage,
  };
};
