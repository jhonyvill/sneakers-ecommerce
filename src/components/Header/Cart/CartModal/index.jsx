import * as S from "./styles";
import { useContext, useEffect } from "react";
import { CartContext } from "../../../../contexts/CartContext";
import { useCart } from "../../../../hooks/useCart";
import CartItem from "./CartItem";
import Button from "../../../Utils/Button";

const CartModal = ({ cartRef, closeCartModal }) => {
  const { cartState } = useContext(CartContext);
  const { cartHasItems } = useCart();

  function handleClickOutside(event) {
    if (!cartRef.current.contains(event.target)) closeCartModal();
  }

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);
    return () => document.removeEventListener("mousedown", handleClickOutside);
  }, []);

  return (
    <S.Modal>
      <S.Title>
        <h2>Cart</h2>
      </S.Title>
      <S.Content>
        {cartHasItems ? (
          <>
            <S.Items>
              {cartState.items.map((item, index) => (
                <li key={index}>
                  <CartItem item={item} />
                </li>
              ))}
            </S.Items>
            <Button>Checkout</Button>
          </>
        ) : (
          <p>Your cart is empty.</p>
        )}
      </S.Content>
    </S.Modal>
  );
};

export default CartModal;
