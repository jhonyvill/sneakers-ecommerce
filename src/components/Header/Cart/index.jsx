import React, { useRef, useState } from "react";
import * as S from "./styles";
import CartSvg from "../../../assets/icon-cart.svg?react";
import CartModal from "./CartModal";
import { useCart } from "../../../hooks/useCart";

const Cart = () => {
  const [isCartModalOpen, setIsCartModalOpen] = useState(false);
  const cartRef = useRef(null);
  const { totalQuantityItemsCart, cartHasItems } = useCart();

  function handleCartModal() {
    setIsCartModalOpen(!isCartModalOpen);
  }

  return (
    <div ref={cartRef}>
      <S.Cart
        aria-label="carrinho de compras"
        aria-expanded={isCartModalOpen}
        onClick={handleCartModal}
      >
        <CartSvg />
        {cartHasItems && (
          <S.CartDetail>{totalQuantityItemsCart}</S.CartDetail>
        )}
      </S.Cart>
      {isCartModalOpen && (
        <CartModal
          cartRef={cartRef}
          closeCartModal={() => setIsCartModalOpen(false)}
          hasItems={cartHasItems}
        />
      )}
    </div>
  );
};

export default Cart;
