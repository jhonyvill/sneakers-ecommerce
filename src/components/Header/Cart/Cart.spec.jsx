import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithTheme } from "../../../../.jest/utils";
import * as hookCart from "../../../hooks/useCart";
import { CartProvider } from "../../../contexts/CartContext";
import Cart from ".";

function renderCart() {
  return renderWithTheme(
    <CartProvider>
      <Cart />
    </CartProvider>
  );
}

function createUseCartSpy(hasItem, quantityItems) {
  return jest.spyOn(hookCart, "useCart").mockReturnValue({
    clearCart: jest.fn(),
    cartHasItems: hasItem,
    totalQuantityItemsCart: quantityItems,
  });
}

describe("<Cart />", () => {
  it("should render Cart component", () => {
    createUseCartSpy(false, 0);
    renderCart();

    expect(screen.getByRole("button")).toBeInTheDocument();
    expect(screen.getByTestId("cart-svg")).toBeInTheDocument();
    expect(screen.queryByTestId("cart-detail")).not.toBeInTheDocument();
    expect(screen.getByRole("region", { name: "carrinho de compras" })).toBeInTheDocument();
  });
  it("should render Cart component with quantity items detail", () => {
    const spy = createUseCartSpy(true, 2);
    renderCart();
    const cartDetail = screen.getByTestId("cart-detail");

    expect(cartDetail).toBeInTheDocument();
    expect(cartDetail).toHaveTextContent("2");
    spy.mockRestore();
  });
  it("should open cart modal when cart button clicked", async () => {
    const spy = createUseCartSpy(false, 0);
    renderCart();
    const button = screen.getByRole("button");

    await userEvent.click(button);

    expect.assertions(2);
    expect(screen.getByRole("region", { name: "modal carrinho de compras" })).toBeInTheDocument();
    expect(screen.getByText(/your cart is empty\./i)).toBeInTheDocument();
    spy.mockRestore();
  });
});

describe("<Cart /> snapshots", () => {
  it("should match snapshot when cart is empty", () => {
    createUseCartSpy(false, 0);
    const { container } = renderCart();
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should match snapshot when cart not is empty", () => {
    createUseCartSpy(true, 1);
    const { container } = renderCart();
    expect(container.firstChild).toMatchSnapshot();
  });
  it("should match snapshot when cart modal open", async () => {
    createUseCartSpy(true, 1);
    const { container } = renderCart();
    const button = screen.getByRole("button");

    await userEvent.click(button);
    expect(container.firstChild).toMatchSnapshot();
  });
});
