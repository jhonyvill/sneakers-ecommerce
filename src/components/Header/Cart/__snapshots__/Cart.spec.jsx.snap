// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> snapshots should match snapshot when cart is empty 1`] = `
<section
  aria-label="carrinho de compras"
>
  <button
    aria-expanded="false"
    class="styles__Cart-sc-33q7z1-0 cfnkFh"
  >
    <icon-mock
      data-testid="cart-svg"
    />
  </button>
</section>
`;

exports[`<Cart /> snapshots should match snapshot when cart modal open 1`] = `
<section
  aria-label="carrinho de compras"
>
  <button
    aria-expanded="true"
    class="styles__Cart-sc-33q7z1-0 cfnkFh"
  >
    <icon-mock
      data-testid="cart-svg"
    />
    <span
      class="styles__CartDetail-sc-33q7z1-1 gvvFRN"
      data-testid="cart-detail"
    >
      1
    </span>
  </button>
  <section
    aria-label="modal carrinho de compras"
    class="styles__Modal-sc-1muh66y-0 joXRfi"
  >
    <div
      class="styles__Title-sc-1muh66y-1 mEFUd"
    >
      <h2>
        Cart
      </h2>
    </div>
    <div
      class="styles__Content-sc-1muh66y-2 bUNORX"
    >
      <ul
        class="styles__Items-sc-1muh66y-3 jSlPFn"
      />
      <button
        aria-label="finalizar compra"
        class="styles__StyledButton-sc-1csa66j-0 hmYbLP"
      >
        Checkout
      </button>
    </div>
  </section>
</section>
`;

exports[`<Cart /> snapshots should match snapshot when cart not is empty 1`] = `
<section
  aria-label="carrinho de compras"
>
  <button
    aria-expanded="false"
    class="styles__Cart-sc-33q7z1-0 cfnkFh"
  >
    <icon-mock
      data-testid="cart-svg"
    />
    <span
      class="styles__CartDetail-sc-33q7z1-1 gvvFRN"
      data-testid="cart-detail"
    >
      1
    </span>
  </button>
</section>
`;
